File: cmd/dependabot-directories/merge_test.go
package main

// Ports of the bats coverage, refusal and template tests, plus the three
// yaml.v3-era tests that replace the two bats tests of the deleted YAML
// backend shim.

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// merge mirrors the bats helper: detect a tree and merge against an existing
// config, in --dry-run mode.
func merge(t *testing.T, tree, cfg string, extra ...string) (int, string) {
	t.Helper()
	args := append([]string{
		"acme/widgets",
		"--paths-from-file", "testdata/trees/" + tree + ".paths",
		"--existing-config", "testdata/dependabot/" + cfg,
		"--dry-run",
	}, extra...)
	return execute(t, args...)
}

func refuteContains(t *testing.T, output, fragment string) {
	t.Helper()
	if strings.Contains(output, fragment) {
		t.Errorf("output must not contain %q\noutput:\n%s", fragment, output)
	}
}

// --- coverage of an existing config -----------------------------------------

func TestAFullyCoveringSingularEntryProducesNoChanges(t *testing.T) {
	code, out := merge(t, "composer-single", "covered-singular.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, "already covers every detected directory")
	assertContains(t, out, "0 would change")
}

func TestAnExistingGlobCoversTheDirectoriesItExpandsTo(t *testing.T) {
	code, out := merge(t, "composer-siblings", "covered-glob.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, "already mapped")
	refuteContains(t, out, "+  - package-ecosystem")
}

func TestATargetBranchEntryDoesNotCountAsCoverage(t *testing.T) {
	// Dependabot's security updates ignore target-branch entries, so
	// counting one as coverage would reintroduce the very bug this tool
	// exists to fix.
	code, out := merge(t, "composer-single", "target-branch.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+  - package-ecosystem: "composer"`)
}

func TestPartialCoverageAppendsOnlyTheGapAndDegradesTheGlob(t *testing.T) {
	code, out := merge(t, "composer-siblings", "partial-commented.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, "would overlap")
	assertContains(t, out, `directory: "/projects/plugins/b"`)
	assertContains(t, out, `directory: "/projects/plugins/c"`)
	// Overlapping entries are what make Dependabot reject a config outright.
	refuteContains(t, out, `"/projects/plugins/*"`)
}

func TestAnExistingEntryWeDetectedNothingForIsReportedNotRemoved(t *testing.T) {
	code, out := merge(t, "composer-single", "stale-entry.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, "matches nothing we detected")
	refuteContains(t, out, "-  - package-ecosystem")
}

func TestCommentsAndKeyOrderOutsideTheAppendedRegionArePreserved(t *testing.T) {
	code, out := merge(t, "composer-siblings", "partial-commented.yml")
	assertExit(t, code, 0, out)
	// An append-only text splice: nothing in the original may show as
	// removed.
	refuteContains(t, out, "-# Managed by the platform team")
	refuteContains(t, out, "-  # the first plugin only")
}

func TestTheFilesOwnIndentationIsMatchedNotOurs(t *testing.T) {
	code, out := merge(t, "composer-single", "four-space-indent.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+    - package-ecosystem: "composer"`)
	assertContains(t, out, `+      directory: "/"`)
}

func TestAnUpdatesKeyWithNoItemsIsAValidAppendTarget(t *testing.T) {
	code, out := merge(t, "composer-single", "empty-updates.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+  - package-ecosystem: "composer"`)
}

func TestAFollowingTopLevelKeyIsNotSwallowedByTheAppend(t *testing.T) {
	code, out := merge(t, "composer-single", "registries-after.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+  - package-ecosystem: "composer"`)
	refuteContains(t, out, "-registries:")
}

// --- refusals ---------------------------------------------------------------

func TestInlineUpdatesFormIsRefused(t *testing.T) {
	code, out := merge(t, "composer-single", "inline-updates.yml")
	assertExit(t, code, 1, out)
	assertContains(t, out, "inline form")
}

func TestTabIndentationIsRefused(t *testing.T) {
	code, out := merge(t, "composer-single", "tab-indent.yml")
	assertExit(t, code, 1, out)
	assertContains(t, out, "tabs")
}

func TestAnchorsAndAliasesAreRefused(t *testing.T) {
	code, out := merge(t, "composer-single", "anchors.yml")
	assertExit(t, code, 1, out)
	assertContains(t, out, "anchors")
}

func TestAnUnparseableConfigStopsTheRunRatherThanAppendingBlindly(t *testing.T) {
	broken := filepath.Join(t.TempDir(), "broken.yml")
	if err := os.WriteFile(broken, []byte("version: 2\nupdates:\n  - [oops\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	code, out := execute(t, "acme/widgets",
		"--paths-from-file", "testdata/trees/composer-single.paths",
		"--existing-config", broken, "--dry-run")
	assertExit(t, code, 1, out)
	// Never fall through to "nothing is covered, append everything".
	refuteContains(t, out, "+  - package-ecosystem")
}

// --- entry template ---------------------------------------------------------

func TestEntriesDefaultToSecurityOnly(t *testing.T) {
	_, out := merge(t, "composer-single", "empty-updates.yml")
	assertContains(t, out, "open-pull-requests-limit: 0")
	refuteContains(t, out, "groups:")
}

func TestEnableVersionUpdatesSwapsInTheFullHouseTemplate(t *testing.T) {
	_, out := merge(t, "composer-single", "empty-updates.yml", "--enable-version-updates")
	assertContains(t, out, "open-pull-requests-limit: 10")
	assertContains(t, out, "composer-minor-patch:")
	assertContains(t, out, "composer-major:")
}

func TestGeneratedEntriesCarryTheHouseCooldown(t *testing.T) {
	_, out := merge(t, "composer-single", "empty-updates.yml")
	assertContains(t, out, "cooldown:")
	assertContains(t, out, "default-days: 7")
}

// --- yaml.v3-era replacements for the backend-shim tests --------------------

func TestCRLFLineEndingsSurviveTheSpliceByteForByte(t *testing.T) {
	// The splice splits on \n alone, so the \r bytes of a CRLF file ride
	// along unchanged — the diff's context lines prove it.
	code, out := merge(t, "composer-single", "crlf.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+  - package-ecosystem: "composer"`)
	assertContains(t, out, "directory: \"/somewhere/else\"\r\n")
	assertContains(t, out, "matches nothing we detected")
}

func TestAMissingTrailingNewlineIsNormalizedNotDuplicated(t *testing.T) {
	code, out := merge(t, "composer-single", "no-trailing-newline.yml")
	assertExit(t, code, 0, out)
	assertContains(t, out, `+  - package-ecosystem: "composer"`)
	// Both diff inputs end in exactly one newline, so diff never marks one.
	refuteContains(t, out, "No newline")
}

func TestDuplicateMappingKeysAreRefused(t *testing.T) {
	// The bash shim's backends silently took the last value; yaml.v3
	// refuses, which is the fail-closed direction the spec demands.
	dup := filepath.Join(t.TempDir(), "dup.yml")
	if err := os.WriteFile(dup, []byte("version: 2\nversion: 3\nupdates:\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	code, out := execute(t, "acme/widgets",
		"--paths-from-file", "testdata/trees/composer-single.paths",
		"--existing-config", dup, "--dry-run")
	assertExit(t, code, 1, out)
	assertContains(t, out, "cannot parse")
	refuteContains(t, out, "+  - package-ecosystem")
}

File: tests/Feature/MergeTest.php
<?php

declare(strict_types=1);

/*
 * Ports of the bats coverage, refusal and template tests, plus the three
 * parser-era tests that replaced the two bats tests of the deleted YAML
 * backend shim.
 */

// --- coverage of an existing config -----------------------------------------

it('produces no changes for a fully covering singular entry', function () {
    [$code, $out] = mergeFixture('composer-single', 'covered-singular.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('already covers every detected directory')
        ->toContain('0 would change');
});

it('treats an existing glob as covering the directories it expands to', function () {
    [$code, $out] = mergeFixture('composer-siblings', 'covered-glob.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('already mapped')->not->toContain('+  - package-ecosystem');
});

it('does not count a target-branch entry as coverage', function () {
    // Dependabot's security updates ignore target-branch entries, so counting
    // one as coverage would reintroduce the very bug this tool exists to fix.
    [$code, $out] = mergeFixture('composer-single', 'target-branch.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+  - package-ecosystem: "composer"');
});

it('appends only the gap and degrades the glob on partial coverage', function () {
    [$code, $out] = mergeFixture('composer-siblings', 'partial-commented.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('would overlap')
        ->toContain('directory: "/projects/plugins/b"')
        ->toContain('directory: "/projects/plugins/c"')
        // Overlapping entries are what make Dependabot reject a config
        // outright.
        ->not->toContain('"/projects/plugins/*"');
});

it('reports an existing entry we detected nothing for rather than removing it', function () {
    [$code, $out] = mergeFixture('composer-single', 'stale-entry.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('matches nothing we detected')->not->toContain('-  - package-ecosystem');
});

it('preserves comments and key order outside the appended region', function () {
    [$code, $out] = mergeFixture('composer-siblings', 'partial-commented.yml');

    expect($code)->toBe(0);
    // An append-only text splice: nothing in the original may show as removed.
    expect($out)->not->toContain('-# Managed by the platform team')
        ->not->toContain('-  # the first plugin only');
});

it("matches the file's own indentation rather than ours", function () {
    [$code, $out] = mergeFixture('composer-single', 'four-space-indent.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+    - package-ecosystem: "composer"')
        ->toContain('+      directory: "/"');
});

it('treats an updates key with no items as a valid append target', function () {
    [$code, $out] = mergeFixture('composer-single', 'empty-updates.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+  - package-ecosystem: "composer"');
});

it('does not swallow a following top-level key', function () {
    [$code, $out] = mergeFixture('composer-single', 'registries-after.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+  - package-ecosystem: "composer"')->not->toContain('-registries:');
});

// --- refusals ---------------------------------------------------------------

it('refuses a structure it cannot reason about', function (string $config, string $expected) {
    [$code, $out] = mergeFixture('composer-single', $config);

    expect($code)->toBe(1);
    expect($out)->toContain($expected);
})->with([
    'inline updates form' => ['inline-updates.yml', 'inline form'],
    'tab indentation' => ['tab-indent.yml', 'tabs'],
    'anchors and aliases' => ['anchors.yml', 'anchors'],
]);

it('stops on an unparseable config rather than appending blindly', function () {
    $broken = sys_get_temp_dir().'/broken-'.bin2hex(random_bytes(6)).'.yml';
    file_put_contents($broken, "version: 2\nupdates:\n  - [oops\n");

    [$code, $out] = runTool([
        'acme/widgets',
        '--paths-from-file', fixturePath('trees', 'composer-single.paths'),
        '--existing-config', $broken,
        '--dry-run',
    ]);

    expect($code)->toBe(1);
    // Never fall through to "nothing is covered, append everything".
    expect($out)->not->toContain('+  - package-ecosystem');

    unlink($broken);
});

// --- entry template ---------------------------------------------------------

it('defaults entries to security-only', function () {
    [, $out] = mergeFixture('composer-single', 'empty-updates.yml');

    expect($out)->toContain('open-pull-requests-limit: 0')->not->toContain('groups:');
});

it('swaps in the full house template with --enable-version-updates', function () {
    [, $out] = mergeFixture('composer-single', 'empty-updates.yml', ['--enable-version-updates']);

    expect($out)->toContain('open-pull-requests-limit: 10')
        ->toContain('composer-minor-patch:')
        ->toContain('composer-major:');
});

it('carries the house cooldown on generated entries', function () {
    [, $out] = mergeFixture('composer-single', 'empty-updates.yml');

    expect($out)->toContain('cooldown:')->toContain('default-days: 7');
});

// --- byte-level splice behaviour --------------------------------------------

it('carries CRLF line endings through the splice byte for byte', function () {
    // The splice splits on \n alone, so the \r bytes of a CRLF file ride
    // along unchanged — the diff's context lines prove it.
    [$code, $out] = mergeFixture('composer-single', 'crlf.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+  - package-ecosystem: "composer"')
        ->toContain("directory: \"/somewhere/else\"\r\n")
        ->toContain('matches nothing we detected');
});

it('normalizes a missing trailing newline rather than duplicating it', function () {
    [$code, $out] = mergeFixture('composer-single', 'no-trailing-newline.yml');

    expect($code)->toBe(0);
    expect($out)->toContain('+  - package-ecosystem: "composer"')
        // Both diff inputs end in exactly one newline, so diff never marks one.
        ->not->toContain('No newline');
});

it('refuses duplicate mapping keys', function () {
    // The bash shim's backends silently took the last value. Both the Go
    // build and Symfony Yaml refuse, which is the fail-closed direction the
    // spec demands.
    $duplicate = sys_get_temp_dir().'/dup-'.bin2hex(random_bytes(6)).'.yml';
    file_put_contents($duplicate, "version: 2\nversion: 3\nupdates:\n");

    [$code, $out] = runTool([
        'acme/widgets',
        '--paths-from-file', fixturePath('trees', 'composer-single.paths'),
        '--existing-config', $duplicate,
        '--dry-run',
    ]);

    expect($code)->toBe(1);
    expect($out)->toContain('cannot parse')->not->toContain('+  - package-ecosystem');

    unlink($duplicate);
});
